addons/stock/tests/test_report.py

```diff
@@ -1687,3 +1658,71 @@ def test_report_reception_6_backorders(self):
         backorder.button_validate()
         for move in delivery.move_ids_without_package:
             self.assertEqual(move.state, 'assigned', "All delivery moves should be fully reserved now")
+
+    def test_report_reception_7_done_receipt(self):
+        """ Check the complicated use cases of correct move splitting when assigning when:
+        1. Outgoing qty is greater than incoming qty + total outgoing qty is already reserved
+        2. Receipt is already done and then assigned
+        """
+
+        incoming_qty = 4
+        outgoing_qty = 10
+        self.env['stock.quant'].with_context(inventory_mode=True).create({
+            'product_id': self.product.id,
+            'location_id': self.stock_location.id,
+            'inventory_quantity': outgoing_qty
+        }).action_apply_inventory()
+
+        # create delivery + receipt
+        delivery_form = Form(self.env['stock.picking'], view='stock.view_picking_form')
+        delivery_form.picking_type_id = self.picking_type_out
+        with delivery_form.move_ids_without_package.new() as move_line:
+            move_line.product_id = self.product
+            move_line.product_uom_qty = outgoing_qty
+        delivery = delivery_form.save()
+        delivery.action_confirm()
+
+        receipt_form = Form(self.env['stock.picking'], view='stock.view_picking_form')
+        receipt_form.partner_id = self.partner
+        receipt_form.picking_type_id = self.picking_type_in
+        with receipt_form.move_ids_without_package.new() as move_line:
+            move_line.product_id = self.product
+            move_line.product_uom_qty = incoming_qty
+        receipt = receipt_form.save()
+        receipt.action_confirm()
+        receipt.button_validate()
+
+        self.assertEqual(len(delivery.move_ids_without_package), 1)
+        self.assertEqual(delivery.move_ids_without_package.quantity, outgoing_qty, "Delivery move should already be reserved")
+        report = self.env['report.stock.report_reception']
+
+        # -------------------
+        # check report assign
+        # -------------------
+        report.action_assign(delivery.move_ids_without_package.ids, [incoming_qty], receipt.move_ids_without_package.ids)
+        mto_move = delivery.move_ids_without_package.filtered(lambda m: m.procure_method == 'make_to_order')
+        non_mto_move = delivery.move_ids_without_package - mto_move
+
+        # check that delivery move splits correctly when receipt move is assigned to it, done receipt = can be assigned to reserved outs
+        self.assertEqual(len(delivery.move_ids_without_package), 2, "Delivery moves should have split into assigned + not assigned")
+        self.assertEqual(len(delivery.move_ids_without_package.mapped('move_orig_ids')), 1, "Only 1 delivery + 1 receipt move should be assigned")
+        self.assertEqual(len(receipt.move_ids_without_package.mapped('move_dest_ids')), 1, "Receipt move should remain unsplit")
```

Comment:
The `mapped` is not needed and I find the comment "*Only 1 delivery + 1 receipt move should be assigned*" not very helpful (I let a suggestion but feel free to not follow it)
```suggestion
        self.assertEqual(len(delivery.move_ids_without_package), 2, "Delivery moves should have split into assigned + not assigned")
        self.assertEqual(len(delivery.move_ids_without_package.move_orig_ids), 1, "Delivery move should be linked to 1 assigned receipt move")
        self.assertEqual(len(receipt.move_ids_without_package.move_dest_ids), 1, "Receipt move should remain unsplit")
```

But I'm not sure, maybe checking records' values (qty and move's origin) instead is better to ensure the split was correctly done ?
```suggestion
        self.assertRecordValues(delivery.move_ids_without_package, [
            {'product_qty': incoming_qty, 'move_orig_ids': receipt.move_ids_without_package.ids},
            {'product_qty': outgoing_qty - incoming_qty, 'move_orig_ids': []},
        ])
        self.assertEqual(len(receipt.move_ids_without_package.move_dest_ids), 1, "Receipt move should remain unsplit")
```

Reply:
Whoops, too fast with my copy paste for the `mapped` part (pushed this change already)

For your other suggestions, I don't think they're necessary. There are 2 delivery moves after the `action_assign` is called because it splits it between the linked qty and the not linked qty and then I check those values in the later asserts. I would prefer to avoid the test failing due to the order of the records changing since that part doesn't matter.